src/cargo/sources/git/utils.rs

```diff
@@ -967,6 +1018,7 @@ pub fn fetch(
                         // folder before writing files into it, or else not even open a directory as git repository (which is
                         // also handled here).
                         && err.is_corrupted()
+                        || has_shallow_lock_file(&err)
```

Comment:
For this, there is no test though - I encountered it while testing manually.
This is quite specific and a test can probably be created to validate this kicks in using something like `touch <repo>/.git/shallow.lock`.

Reply:
A test was added now.